### ISSUE TITLE
Support for vhost@connectorname syntax of virtual hosts

### ISSUE THREAD

**Issue by backinside** [eclipse/jetty.project#1785]
After upgrading from jetty8, we can't set the connector name and the virtual host for a context
the setConnectorNames was removed from the WebAppContext
We would like to have multiple apps responding on https, but  other apps responding on http
Is there a way, to achieve this?

**Comment by joakime** [comment]
use setVirtualHost() with "@Name" syntax.

Aka, if your Connector has the name "webadm" then the virtualhost string would be "@webadm"

**Comment by backinside** [comment]
But that means, that the context will capture all the requests, regardless of the virtual host
We would like to have multiple virtual hosts on the same connector

**Comment by joakime** [comment]
@backinside the implementation of [`ContextHandler#checkVirtualHost(Request)`](https://github.com/eclipse/jetty.project/blob/jetty-9.4.6.v20170531/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java#L968) (which does the Virtual Host selection of contexts) already handles @Name + virtualhost combined matching during its check.

It should be safe to use them both in your WebAppContext and it should match on both the @Name and the hostnames.

Or is this not working for you? and you are reporting a bug with it?

**Comment by backinside** [comment]
@joakime Not really, I have this in my context:
        <Set name="virtualHosts">
                <Array type="java.lang.String">
                        <Item>@https</Item>
                        <Item>vhost1</Item>
                </Array>
        </Set>
But it responds to other virtual hosts as well
Meanwhile if I remove the connector, it only responds to vhost1

**Comment by joakime** [comment]
@gregw could this be a virtualHost **"OR"** behavior vs the OP's requested **"AND"** behavior with regards to virtualhosts checking?

Said another way...

A Context with `virtualHosts[] { "@https", "vhost1" };` ...

The `ContextHandler#checkVirtualHost(Request)` currently perform an **"OR"** check (if name _or_ vhost matches its ok to use that context)

But the OP seems to want an **"AND"** check (if name _and_ vhost matches, then its ok to use that context)

**Comment by backinside** [comment]
@joakime to handle all of our theoretical scenarios, maybe handle `virtualHosts[] { "@https+vhost1" };`
this way multiple connection and virtual host combinations can be defined?

**Comment by gregw** [comment]
Perhaps we should support "vhost1@https", so a vhost and connector can be
matched in a single pattern?

**Comment by spbolton** [comment]
I had this same issue myself this week and was thinking of the best way of solving this after finding the cause in the code.  I like the idea of appending with the virtualhost name as this makes it not effect any previous functionality.  I have tried to also clean up this method a bit and added to the unit tests.  This would be my first commit to Jetty,  I have tried to follow the guidelines but was not sure which branch I should create a pull request for, or create a feature branch.  Did not realize the issue number would add it from my clone above anyway.  Thanks
